File: rules-interpreter/src/main/java/com/github/iamrenny/ruleflow/evaluators/ListContextEvaluator.java
package com.github.iamrenny.ruleflow.evaluators;

import com.github.iamrenny.ruleflow.RuleFlowLanguageLexer;
import com.github.iamrenny.ruleflow.RuleFlowLanguageParser;
import com.github.iamrenny.ruleflow.errors.PropertyNotFoundException;
import com.github.iamrenny.ruleflow.visitors.Visitor;

import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class ListContextEvaluator implements ContextEvaluator<RuleFlowLanguageParser.ListContext> {

    @Override
    public Object evaluate(RuleFlowLanguageParser.ListContext ctx, Visitor visitor) throws PropertyNotFoundException {
        if (ctx.not == null && ctx.op.getType() == RuleFlowLanguageLexer.K_CONTAINS) {
            return evalContains(ctx, visitor);
        } else if (ctx.not != null && ctx.op.getType() == RuleFlowLanguageLexer.K_CONTAINS) {
            return !(Boolean) evalContains(ctx, visitor);
        } else if (ctx.not == null && ctx.op.getType() == RuleFlowLanguageLexer.K_IN) {
            return evalIn(ctx, visitor);
        } else if (ctx.not != null && ctx.op.getType() == RuleFlowLanguageLexer.K_IN) {
            return !(Boolean) evalIn(ctx, visitor);
        } else if (ctx.not == null && ctx.op.getType() == RuleFlowLanguageLexer.K_STARTS_WITH) {
            return evalStartsWith(ctx, visitor);
        } else if (ctx.not != null && ctx.op.getType() == RuleFlowLanguageLexer.K_STARTS_WITH) {
            return !(Boolean) evalStartsWith(ctx, visitor);
        } else {
            throw new RuntimeException("Unexpected token near " + ctx.value.getText());
        }
    }

    private Object evalIn(RuleFlowLanguageParser.ListContext ctx, Visitor visitor) throws PropertyNotFoundException {
        Object value = visitor.visit(ctx.value);

        if (ctx.values.literalList != null) {
            List<String> literals = ctx.values.string_literal().stream()
                .map(literal -> literal.getText().replace("'", ""))
                .toList();
            return literals.contains(value);
        } else if (ctx.values.storedList != null) {
            String listKey = ctx.values.string_literal(0).getText().replace("'", "");
            Set<String> list = visitor.getLists().get(listKey);
            return list != null && list.contains(value.toString());
        } else if (ctx.values.validProperty() != null) {
            List<?> validPropertyList = (List<?>) new ValidPropertyContextEvaluator().evaluate(ctx.values.validProperty(), visitor);
            return validPropertyList.contains(value);
        } else {
            throw new RuntimeException("Cannot find symbol " + ctx.values);
        }
    }

    private Object evalContains(RuleFlowLanguageParser.ListContext ctx, Visitor visitor) {
        Object value = visitor.visit(ctx.value);

        if (ctx.values.literalList != null) {
            List<String> literals = ctx.values.string_literal().stream()
                .map(literal -> literal.getText().replace("'", ""))
                .collect(Collectors.toList());
            return literals.stream().anyMatch(literal -> value.toString().contains(literal));
        } else if (ctx.values.storedList != null) {
            String listKey = ctx.values.string_literal(0).getText().replace("'", "");
            Set<String> list = visitor.getLists().get(listKey);
            return list != null && list.stream().anyMatch(item -> value.toString().contains(item));
        } else {
            throw new RuntimeException("Cannot find symbol");
        }
    }

    private Object evalStartsWith(RuleFlowLanguageParser.ListContext ctx, Visitor visitor) throws PropertyNotFoundException {
        String value = visitor.visit(ctx.value).toString();
        List<?> list;

        if (ctx.values.literalList != null) {
            list = ctx.values.string_literal().stream()
                .map(literal -> literal.getText().replace("'", ""))
                .collect(Collectors.toList());
        } else if (ctx.values.storedList != null) {
            String listKey = ctx.values.string_literal(0).getText().replace("'", "");
            list = visitor.getLists().get(listKey).stream().collect(Collectors.toList());
        } else if (ctx.values.validProperty() != null) {
            list = (List<?>) new ValidPropertyContextEvaluator().evaluate(ctx.values.validProperty(), visitor);
        } else {
            throw new RuntimeException("Unexpected symbol " + ctx.values);
        }

        return list.stream()
            .anyMatch(elem -> value.startsWith(elem.toString()));
    }
}

File: rules-interpreter/src/main/kotlin/com/rappi/analang/evaluators/ListContextEvaluator.kt
package com.rappi.analang.evaluators

import com.rappi.fraud.analang.ANALexer
import com.rappi.fraud.analang.ANAParser
import com.rappi.analang.visitors.Visitor


class ListContextEvaluator : ContextEvaluator<ANAParser.ListContext> {

    override fun evaluate(ctx: ANAParser.ListContext, visitor: Visitor): Any {

        return when {
            ctx.not == null && ctx.op.type == ANALexer.K_CONTAINS -> evalContains(ctx, visitor)
            ctx.not != null && ctx.op.type == ANALexer.K_CONTAINS -> evalContains(ctx, visitor) == false
            ctx.not == null && ctx.op.type == ANALexer.K_IN -> evalIn(ctx, visitor)
            ctx.not != null && ctx.op.type == ANALexer.K_IN -> evalIn(ctx, visitor) == false
            ctx.not == null && ctx.op.type == ANALexer.K_STARTS_WITH -> evalStartsWith(ctx, visitor)
            ctx.not != null && ctx.op.type == ANALexer.K_STARTS_WITH -> evalStartsWith(ctx, visitor) == false
            else -> throw RuntimeException("Unexpected token near ${ctx.value.text}")
        }

    }


    private fun evalIn(ctx: ANAParser.ListContext, visitor: Visitor): Any {
        val value = visitor.visit(ctx.value)
        return when {
            (ctx.values.literalList != null) -> {
                ctx.values.string_literal()
                    .map{ it.text.replace("'", "", true) }
                    .map{ it.replace("'", "") }
                    .contains(value)
            }
            ctx.values.storedList != null -> {

                val list = visitor.lists[ctx.values.string_literal()[0].text.replace("\'", "")]

                list?.contains(value.toString()) ?: false
            }
            ctx.values.validProperty() != null ->
                (ValidPropertyContextEvaluator().evaluate(ctx.values.validProperty(), visitor) as List<*>)
                    .contains(value)
            else -> error("Cannot find symbol ${ctx.values}")
        }
    }

    @Suppress("UNCHECKED_CAST")
    private fun evalContains(ctx: ANAParser.ListContext, visitor: Visitor): Any {
       return when {
            ctx.values.literalList != null  -> {
                val value = visitor.visit(ctx.value)
                val list = ctx.values.string_literal()
                list.any {
                    value.toString().contains(it.text.replace("'", ""), true)
                }
            }
            ctx.values.storedList != null -> {
                val value = visitor.visit(ctx.value)
                val list = visitor.lists[ctx.values.string_literal()[0].text.replace("\'", "")]

                list?.any {
                    value.toString().contains(it, true)
                } ?: false
                }
            else -> error("Cannot find symbol")
            }
        }

    private fun evalStartsWith(ctx: ANAParser.ListContext, visitor: Visitor): Any {
        val value = visitor.visit(ctx.value).toString()
        val list =  when {
            (ctx.values.literalList != null) -> {
                ctx.values.string_literal()
                    .map{it.text.replace("'", "", true)}
            }
            ctx.values.storedList != null -> {
               visitor.lists[ctx.values.string_literal()[0].text.replace("\'", "")]
            }
            ctx.values.validProperty() != null -> (ValidPropertyContextEvaluator().evaluate(ctx.values.validProperty(), visitor) as List<*>)

            else -> error("Unexpected symbol ${ctx.values}")
        }
        return list?.map { elem -> value.startsWith(elem.toString(), true) }
            ?.reduce { fst, snd -> fst || snd } ?: false
    }


}